bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/link/LinkHandler.java

```diff
@@ -0,0 +1,126 @@
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.link;
+
+import static com.adobe.cq.wcm.core.components.internal.link.LinkNameConstants.PN_LINK_TARGET;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkNameConstants.PN_LINK_URL;
+import static com.adobe.cq.wcm.core.components.internal.link.LinkNameConstants.VALID_LINK_TARGETS;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.ScriptVariable;
+import org.apache.sling.models.annotations.injectorspecific.Self;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.adobe.cq.wcm.core.components.internal.Utils;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+
+/**
+ * Simple implementation for resolving and validating links from model's resources.
+ * This is a Sling model that can injected into other models using the <code>@Self</code> annotation.
+ */
+@Model(adaptables=SlingHttpServletRequest.class)
+public class LinkHandler {
+
+    @Self
+    private SlingHttpServletRequest request;
+    @ScriptVariable
+    private PageManager pageManager;
+
+    /**
+     * Resolve a link from the properties of the given resource.
+     * @param resource Resource
+     * @return Link may be invalid, but is never null
+     */
+    public @NotNull Link getLink(@NotNull Resource resource) {
+        return getLink(resource, PN_LINK_URL);
+    }
+
+    /**
+     * Resolve a link from the properties of the given resource.
+     * @param resource Resource
+     * @param linkURLPropertyName Property name to read link URL from.
+     * @return Link may be invalid, but is never null
+     */
+    public @NotNull Link getLink(@NotNull Resource resource, String linkURLPropertyName) {
+        ValueMap props = resource.getValueMap();
+        String linkURL = props.get(linkURLPropertyName, String.class);
+        String linkTarget = props.get(PN_LINK_TARGET, String.class);
+        return getLink(linkURL, linkTarget);
+    }
+
+    /**
+     * Build a link pointing to the given target page.
+     * @param page Target page
+     * @return Link may be invalid, but is never null
+     */
+    @SuppressWarnings("deprecation")
+    public @NotNull Link getLink(@Nullable Page page) {
+        if (page == null) {
+            return getInvalid();
+        }
+        String linkURL = Utils.getURL(request, page);
```

Comment:
I wouldn't use a deprecated API. I would prefer to introduce a private method here to do the job.

Reply:
well, i just deprecated the used methods it in the same commit.
i've not removed the Utils.getURL methods completely, but that would be one of the things i would do next and include the code in the link handler itself. just wanted not to throw everything away at once to make reviewing the PRs easier.

and Utils is only an internal class, not an external.

if you prefer i can get rid of the deprecated methods rightaway in this PR as well.

Reply:
+1